Test/tutorial/induction-principle.md

```diff
@@ -0,0 +1,779 @@
+proof follows the structure of this function. This process is time consuming,
+leads to big proofs which are hard to maintain, and is error prone as we are blind while
+instantiating function calls and lemmas calls (in particular, it is very easy to mix up
+variable indices).
+
+For instance, let's look at the correctness proof for the ``IsPure`` predicate, that we introduce
+below:
+
+```dafny
+predicate method IsPure(e: Expr, locals: set<string> := {})
+  decreases e
+  // `locals`: the set of local variables which have been introduced by outer
+  // let-bindings (if the expression only updates those variables then it is
+  // pure, because those variables don't escape the scope of the term).
+{
+  match e
+    case Var(name) => true
+    case Literal(n) => true
+    case Bind(bvar, bval, body) =>
+      IsPure(bval, locals)
+      && IsPure(body, {bvar} + locals)
+    case Assign(avar, aval) =>
+      avar in locals && IsPure(aval, locals)
+    case If(cond, thn, els) =>
+      && IsPure(cond, locals)
+      && IsPure(thn, locals)
+      && IsPure(els, locals)
+    case Op(op, oe1, oe2) =>
+      IsPure(oe1, locals) && IsPure(oe2, locals)
+    case Seq(e1, e2) =>
+      IsPure(e1, locals) && IsPure(e2, locals)
+}
+```
+
+We want to prove the following theorem:
+
+<a id="interpexpr-pure"></a>
+```dafny
+lemma InterpExpr_Pure(e: Expr, ctx: Context)
+  requires Pure.IsPure(e)
+  ensures
+    match InterpExpr(e, ctx)
+      case Success((_, ctx')) => ctx' == ctx // state is unchanged
+      case Failure => true
+```
+
+where a context is a map from strings to values:
+```dafny
+type Context = map<string, int>
+```
+
+This theorem can be deduced from the following intermediate lemma, that we prove by
+induction on the expression `e`:
+
+```dafny
+lemma InterpExpr_Pure(e: Expr, locals: set<string>, ctx: Context)
+  requires IsPure(e, locals)
+  requires ctx.Keys >= locals
+  ensures
+    match InterpExpr(e, ctx)
+      case Success((_, ctx')) =>
+        && ctx'.Keys == ctx.Keys
+        // We only update variables captured by outer let-bindings
+        && forall x | x in ctx.Keys && x !in locals :: ctx'[x] == ctx[x]
+      case Failure => true
+```
+
+Let's consider the `Seq` case. In our AST, we define sequences as follows:
+
+```dafny
+datatype Expr =
+  | Seq(e1: Expr, e2: Expr)
+  | ... // Omitted
+```
+
+And the interpreter is defined on sequences as follows:
+
+```dafny
+function method InterpExpr(e: Expr, ctx: Context): Result<(int, Context)> {
+  match e
+    case Seq(e1, e2) =>
+      var (_, ctx1) :- InterpExpr(e1, ctx); // Evaluate the first expression
+      InterpExpr(e2, ctx1) // Continue
+
+    // Omitted
+    case ... =>
+      ...
+}
+```
+
+We face the problem that the proof of the `Seq` case requires introducing the same
+intermediate values as the ones manipulated by the interpreter (like `ctx1`), so that we
+can *write* the recursive calls to `InterpExpr_Pure_WithLocals` (which give us our
```

Comment:
What is this lemma? How does it relate to `InterpExpr` ? I'm not sure I understand the logic here. It uses "..." for the spec, but it was never defined before.

Reply:
Thanks for this! I renamed ands merged some predicates just before the PR, and missed the associated lemmas. I updated the document accordingly (here, the reference should be to `InterpExpr_Pure`, which is declared just above, hence the `...`).